internal/db/gc.go

```diff
@@ -0,0 +1,98 @@
+package db
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype"
+)
+
+func GcProjectObjects(ctx context.Context, tx pgx.Tx, project int64, keep int64, fromVersion int64) ([]Hash, error) {
+	hashes := []Hash{}
+
+	rows, err := tx.Query(ctx, `
+		WITH latest AS (
+			SELECT latest_version AS version
+			FROM dl.projects
+			WHERE id = $1
+		)
+		DELETE FROM dl.objects
+		WHERE project = $1
+		  AND start_version > $2
+		  AND stop_version IS NOT NULL
+		  AND stop_version <= ((SELECT version FROM latest) - $3)
+		RETURNING path, (hash).h1, (hash).h2
+	`, project, fromVersion, keep)
+	if err != nil {
+		return hashes, fmt.Errorf("GcProjectObjects query, project %v, keep %v, from %v: %w", project, keep, fromVersion, err)
+	}
+
+	for rows.Next() {
+		var path string
+		var hash Hash
+		err = rows.Scan(&path, &hash.H1, &hash.H2)
+		if err != nil {
+			return hashes, fmt.Errorf("GcProjectObjects scan %v: %w", project, err)
+		}
+
+		hashes = append(hashes, hash)
+	}
+
+	return hashes, nil
+}
+
+func GcContentHashes(ctx context.Context, tx pgx.Tx, hashes []Hash) (int64, error) {
+	if len(hashes) == 0 {
+		return 0, nil
+	}
+
+	// FIXME: Do this on server boot
+	hashType, err := tx.Conn().LoadType(ctx, "hash")
+	if err != nil {
+		return 0, fmt.Errorf("load hash type: %w", err)
+	}
+	tx.Conn().TypeMap().RegisterType(hashType)
+
+	var params strings.Builder
+	for idx := 1; idx <= len(hashes); idx++ {
+		params.WriteString(fmt.Sprintf("$%d::hash", idx))
+		if idx != len(hashes) {
+			params.WriteRune(',')
+		}
+	}
+
+	query := fmt.Sprintf(`
+		WITH missing AS (
+			SELECT c.hash
+			FROM dl.contents c
+			LEFT JOIN dl.objects o
+				ON c.hash = o.hash
+			WHERE c.hash IN (%s)
+			  AND o.hash IS NULL
+		)
+		DELETE FROM dl.contents
+		WHERE hash IN (SELECT hash FROM missing)
+	`, params.String())
+
+	tag, err := tx.Exec(ctx, query, encodeHashes(hashes)...)
+	if err != nil {
+		return 0, fmt.Errorf("GcContentHashes query, hash count %v: %w", len(hashes), err)
+	}
+
+	return tag.RowsAffected(), nil
+}
+
+func encodeHashes(hashes []Hash) []any {
+	fields := make([]any, len(hashes))
+
+	for idx, hash := range hashes {
+		fields[idx] = pgtype.CompositeFields{
+			pgtype.UUID{Bytes: *(*[16]byte)(hash.H1), Valid: true},
+			pgtype.UUID{Bytes: *(*[16]byte)(hash.H2), Valid: true},
```

Comment:
funky 😵‍💫

Reply:
Supposedly Go 1.18 supports slice to byte array conversion `([16]byte)hash.H1` but I couldn't get it to work without this weird pointer indirection...